{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook descrbies the PORT-EK pipeline analysis of k-mers genreated from the \"deer\" dataset from [reference to our paper].\n",
    "To use it you will need the k-mer indices generated with PORTEKfind.py in the output/deer/15mer_indices directory.\n",
    "To generate the indices as in the paper, you will need to run PORTEKfind.py on the appropriate GISAID data sets, previously downloaded in .fasta files, using the following commands from the main PORT-EK directory:\n",
    " - python PORTEKfind.py \"input/deer/EPI_SET_240422va.fasta\" \"output/deer/15mer_indices/\" --k 15 --group deer\n",
    " - python PORTEKfind.py \"input/deer/EPI_SET_240422rw.fasta\" \"output/deer/15mer_indices/\" --k 15 --group humearly  \n",
    " - python PORTEKfind.py \"input/deer/EPI_SET_240422qc.fasta\" \"output/deer/15mer_indices/\" --k 15 --group humlate  \n",
    "\n",
    "For the selection of optimal k please see optimal_deer_k_selection.ipynb notebook. \n",
    "Note that GISAID web interface only allows downloading up to 10 000 sequences in one file, so when you download the 04.2021 and 11.2021 human viral sequences they will be in several files. You have to first combine them into a single .fasta file or run PORTEKfind separately on the partial files. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Import necessary libraries and PORT-EK source code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import json\n",
    "import pathlib\n",
    "import pickle\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "#adding portek source directory to sys path before importing\n",
    "portek_path = \"../portek\"\n",
    "sys.path.insert(0,portek_path)\n",
    "import portek\n",
    "\n",
    "pd.options.mode.copy_on_write = True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Declare data set specific definitions and functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PORT-EK parameters\n",
    "c = 0.01  # This is the conservation thershold used in k-mer rarity filter\n",
    "m = 2  # This is the maximum number of mismatches allowed when re-examining rare k-mers\n",
    "min_RMSE = 0.1  # This is the RMSE threshold used to select enriched k-mers\n",
    "m_map = 2  # This is the maximum number of mismatches allowed when mapping k-mers to reference genome\n",
    "l_map = 1000  # This the maximum allowed offset of mapping position from average position of k-mer in samples\n",
    "\n",
    "#Relative path to k-mer indices\n",
    "INPUT_PATH = \"../output/deer/15mer_indices\"\n",
    "\n",
    "#Data set specific definitions of k-mer type, column names, reference gene and protein mapping, and colors for plots.\n",
    "\n",
    "FREQ_COLS = ['deer_freq','humearly_freq','humlate_freq']\n",
    "AVG_COLS = ['deer_avg','humearly_avg','humlate_avg']\n",
    "\n",
    "VOLCANO_CMAP = {\n",
    "    \"not significant\": (\"#DDDDDD\", 0.5),\n",
    "    \"deer enriched\": (\"#ffa401\", 1),\n",
    "    \"human enriched\": (\"#005ff5\", 1),\n",
    "    \"time-dependant\": (\"#b99e9e\", 0.8),\n",
    "}\n",
    "GENE_ORDER_LIST = [\n",
    "    \"orf1ab\",\n",
    "    \"S\",\n",
    "    \"orf3a\",\n",
    "    \"E\",\n",
    "    \"M\",\n",
    "    \"orf6\",\n",
    "    \"orf7a\",\n",
    "    \"orf7b\",\n",
    "    \"orf8\",\n",
    "    \"N\",\n",
    "    \"orf10\",\n",
    "    \"intergenic\",\n",
    "]\n",
    "PROTEIN_ORDER_LIST = [\n",
    "    \"nsp1\",\n",
    "    \"nsp2\",\n",
    "    \"nsp3\",\n",
    "    \"nsp4\",\n",
    "    \"nsp5\",\n",
    "    \"nsp6\",\n",
    "    \"nsp7\",\n",
    "    \"nsp8\",\n",
    "    \"nsp9\",\n",
    "    \"nsp10\",\n",
    "    \"nsp12\",\n",
    "    \"nsp13\",\n",
    "    \"nsp14\",\n",
    "    \"nsp15\",\n",
    "    \"nsp16\",\n",
    "    \"S\",\n",
    "    \"orf3a\",\n",
    "    \"E\",\n",
    "    \"M\",\n",
    "    \"orf6\",\n",
    "    \"orf7a\",\n",
    "    \"orf7b\",\n",
    "    \"orf8\",\n",
    "    \"N\",\n",
    "    \"orf10\",\n",
    "    \"non-coding\",\n",
    "]\n",
    "\n",
    "\n",
    "def assign_kmer_type(row):\n",
    "    if (\n",
    "        row[\"deer_early_err\"] > 0\n",
    "        and row[\"deer_late_err\"] > 0\n",
    "        and row[\"p-value_late\"] < 0.01\n",
    "        and row[\"p-value_early\"] < 0.01\n",
    "    ):\n",
    "        return \"deer over-represented\"\n",
    "    elif (\n",
    "        row[\"deer_early_err\"] < 0\n",
    "        and row[\"deer_late_err\"] < 0\n",
    "        and row[\"p-value_late\"] < 0.01\n",
    "        and row[\"p-value_early\"] < 0.01\n",
    "    ):\n",
    "        return \"human over-represented\"\n",
    "    elif row[\"p-value_early\"] < 0.01 or row[\"p-value_late\"] < 0.01:\n",
    "        return \"time-dependant\"\n",
    "    else:\n",
    "        return \"not significant\"\n",
    "\n",
    "\n",
    "def assign_gene(pos):\n",
    "    if pos in range(266, 21555 + 1):\n",
    "        gene = \"orf1ab\"\n",
    "    elif pos in range(21563, 25384 + 1):\n",
    "        gene = \"S\"\n",
    "    elif pos in range(25393, 26220 + 1):\n",
    "        gene = \"orf3a\"\n",
    "    elif pos in range(26245, 26472 + 1):\n",
    "        gene = \"E\"\n",
    "    elif pos in range(26523, 27191 + 1):\n",
    "        gene = \"M\"\n",
    "    elif pos in range(27202, 27387 + 1):\n",
    "        gene = \"orf6\"\n",
    "    elif pos in range(27394, 27759 + 1):\n",
    "        gene = \"orf7a\"\n",
    "    elif pos in range(27756, 27887 + 1):\n",
    "        gene = \"orf7b\"\n",
    "    elif pos in range(27894, 28259 + 1):\n",
    "        gene = \"orf8\"\n",
    "    elif pos in range(28274, 29533 + 1):\n",
    "        gene = \"N\"\n",
    "    elif pos in range(29558, 29674 + 1):\n",
    "        gene = \"orf10\"\n",
    "    else:\n",
    "        gene = \"intergenic\"\n",
    "    return gene\n",
    "\n",
    "\n",
    "def assign_protein(pos):\n",
    "    if pos in range(266, 805 + 1):\n",
    "        protein = \"nsp1\"\n",
    "    elif pos in range(806, 2719 + 1):\n",
    "        protein = \"nsp2\"\n",
    "    elif pos in range(2720, 8554 + 1):\n",
    "        protein = \"nsp3\"\n",
    "    elif pos in range(8555, 10054 + 1):\n",
    "        protein = \"nsp4\"\n",
    "    elif pos in range(10055, 10972 + 1):\n",
    "        protein = \"nsp5\"\n",
    "    elif pos in range(10973, 11842 + 1):\n",
    "        protein = \"nsp6\"\n",
    "    elif pos in range(11843, 12091 + 1):\n",
    "        protein = \"nsp7\"\n",
    "    elif pos in range(12092, 12685 + 1):\n",
    "        protein = \"nsp8\"\n",
    "    elif pos in range(12686, 13024 + 1):\n",
    "        protein = \"nsp9\"\n",
    "    elif pos in range(13025, 13441 + 1):\n",
    "        protein = \"nsp10\"\n",
    "    elif pos in range(13442, 16236 + 1):\n",
    "        protein = \"nsp12\"\n",
    "    elif pos in range(16237, 18039 + 1):\n",
    "        protein = \"nsp13\"\n",
    "    elif pos in range(18040, 19620 + 1):\n",
    "        protein = \"nsp14\"\n",
    "    elif pos in range(19621, 20658 + 1):\n",
    "        protein = \"nsp15\"\n",
    "    elif pos in range(20659, 21555 + 1):\n",
    "        protein = \"nsp16\"\n",
    "    elif pos in range(21563, 25384 + 1):\n",
    "        protein = \"S\"\n",
    "    elif pos in range(25393, 26220 + 1):\n",
    "        protein = \"orf3a\"\n",
    "    elif pos in range(26245, 26472 + 1):\n",
    "        protein = \"E\"\n",
    "    elif pos in range(26523, 27191 + 1):\n",
    "        protein = \"M\"\n",
    "    elif pos in range(27202, 27387 + 1):\n",
    "        protein = \"orf6\"\n",
    "    elif pos in range(27394, 27759 + 1):\n",
    "        protein = \"orf7a\"\n",
    "    elif pos in range(27756, 27887 + 1):\n",
    "        protein = \"orf7b\"\n",
    "    elif pos in range(27894, 28259 + 1):\n",
    "        protein = \"orf8\"\n",
    "    elif pos in range(28274, 29533 + 1):\n",
    "        protein = \"N\"\n",
    "    elif pos in range(29558, 29674 + 1):\n",
    "        protein = \"orf10\"\n",
    "    else:\n",
    "        protein = \"non-coding\"\n",
    "    return protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Construct k-mer count matrix and apply rarity filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Imported 371342 kmers and 33767 samples.\n",
      "Completed humearly_EPI_ISL_1636116. 33767 of 33767 indices done..\n",
      "41946 kmers remaining after filtering at a threshold of 0.01.\n",
      "Finished importing count matrices and calculating frequencies.\n",
      "Total matrix size is 371342 kmers by 33767 samples.\n"
     ]
    }
   ],
   "source": [
    "kmer_set = set()\n",
    "sample_list = []\n",
    "in_path = pathlib.Path(INPUT_PATH).glob('**/*')\n",
    "\n",
    "for filename in in_path:\n",
    "    sample_list.append(filename.stem)\n",
    "    with open(filename, mode=\"r\") as in_file:\n",
    "        temp_dict = json.load(in_file)\n",
    "    kmer_set.update(temp_dict.keys())\n",
    "\n",
    "all_kmer_matrix = pd.DataFrame(0, index=list(kmer_set), columns=sample_list, dtype=\"uint8\")\n",
    "deer_sample_idx = [sample for sample in all_kmer_matrix.columns if \"deer\" in sample]\n",
    "humearly_sample_idx = [sample for sample in all_kmer_matrix.columns if \"humearly\" in sample]\n",
    "humlate_sample_idx = [sample for sample in all_kmer_matrix.columns if \"humlate\" in sample]\n",
    "\n",
    "print(f\"\\nImported {len(kmer_set)} kmers and {len(sample_list)} samples.\")\n",
    "\n",
    "counter = 1\n",
    "tot_files = len(sample_list)\n",
    "in_path = pathlib.Path(INPUT_PATH).glob('**/*')\n",
    "\n",
    "for filename in in_path:\n",
    "    with open(filename, mode=\"r\") as in_file:\n",
    "        temp_dict = json.load(in_file)\n",
    "    count_dict = {f\"{filename.stem}\":[len(pos) for pos in temp_dict.values()]}\n",
    "    temp_df = pd.DataFrame(count_dict, index=temp_dict.keys(), dtype=\"uint8\")\n",
    "    all_kmer_matrix.update(temp_df)\n",
    "    print(f\"Completed {filename.stem}. {counter} of {tot_files} indices done.\", end=\"\\r\", flush=True)\n",
    "    counter += 1\n",
    "\n",
    "bin_kmer_matrix = all_kmer_matrix > 0\n",
    "all_kmer_matrix['deer_freq'] = bin_kmer_matrix.loc[:,deer_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humearly_freq'] = bin_kmer_matrix.loc[:,humearly_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humlate_freq'] = bin_kmer_matrix.loc[:,humlate_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['deer_avg'] = all_kmer_matrix.loc[:,deer_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humearly_avg'] = all_kmer_matrix.loc[:,humearly_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humlate_avg'] = all_kmer_matrix.loc[:,humlate_sample_idx].mean(axis=1)\n",
    "del bin_kmer_matrix\n",
    "\n",
    "if 'AAAAAAAAAAAAAAA' in all_kmer_matrix.index:\n",
    "    all_kmer_matrix = all_kmer_matrix.drop('AAAAAAAAAAAAAAA')\n",
    "\n",
    "common_kmer_matrix = portek.filter_kmers(all_kmer_matrix, freq_cols=FREQ_COLS, cons_thr=c)\n",
    "\n",
    "print(f\"\\n{len(common_kmer_matrix)} common k-mers remaining after filtering at a threshold of {c}.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Calculate common k-mer statistics and identify host over-represented k-mers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "common_kmer_matrix['deer_early_err'] = common_kmer_matrix['deer_avg']-common_kmer_matrix['humearly_avg']\n",
    "common_kmer_matrix['deer_late_err'] = common_kmer_matrix['deer_avg']-common_kmer_matrix['humlate_avg']\n",
    "common_kmer_matrix['deer_RMSE'] =  np.sqrt(((common_kmer_matrix['deer_early_err'])**2+(common_kmer_matrix['deer_late_err'])**2)/2)\n",
    "common_kmer_matrix['seq'] = common_kmer_matrix.index\n",
    "common_kmer_matrix['p-value_early'] = common_kmer_matrix['seq'].apply(portek.calc_kmer_pvalue,args=(deer_sample_idx, humearly_sample_idx, common_kmer_matrix))\n",
    "common_kmer_matrix['p-value_late'] = common_kmer_matrix['seq'].apply(portek.calc_kmer_pvalue,args=(deer_sample_idx, humlate_sample_idx, common_kmer_matrix))\n",
    "common_kmer_matrix = common_kmer_matrix.sort_values('deer_RMSE',ascending=False)\n",
    "common_kmer_matrix = common_kmer_matrix.drop('seq', axis=1)\n",
    "common_kmer_matrix['-log10_p-value_early'] = -np.log10(common_kmer_matrix['p-value_early'])\n",
    "common_kmer_matrix['-log10_p-value_late'] = -np.log10(common_kmer_matrix['p-value_late'])\n",
    "common_kmer_matrix['group'] = common_kmer_matrix.apply(assign_kmer_type, axis=1)\n",
    "common_kmer_stat_matrix = common_kmer_matrix.drop(sample_list, axis=1)\n",
    "deer_overrep = common_kmer_matrix[common_kmer_matrix['group'] == 'deer over-represented']\n",
    "human_overrep = common_kmer_matrix[common_kmer_matrix['group'] == 'human over-represented']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Optional: save/load k-mer count and statistics matrices. Uncomment the relevant lines to save/load particular files."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# common_kmer_matrix.to_csv(\"output/deer/common_15mer_count_matrix.csv\")\n",
    "# common_kmer_stat_matrix.to_csv(\"output/deer/common_15mer_stat_matrix.csv\")\n",
    "# all_kmer_matrix.to_csv(\"output/deer/all_15mer_count_matrix.csv\")\n",
    "\n",
    "# common_kmer_matrix = pd.read_csv('output/deer/common_15mer_count_matrix.csv', index_col=0)\n",
    "# all_kmer_matrix = pd.read_csv('output/deer/all_15mer_count_matrix.csv', index_col=0)\n",
    "# common_kmer_stat_matrix = pd.read_csv(\"output/deer/common_15mer_stat_matrix.csv\", index_col=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Visualize common k-mers using volcano plots. Uncomment the last line in each cell to save as svg."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "sns.scatterplot(data=common_kmer_matrix, x='deer_early_err', y = '-log10_p-value_early', s=10, linewidth = 0, hue='group', palette=VOLCANO_CMAP)\n",
    "plt.axhline(y=-np.log10(0.01), color = 'black')\n",
    "ax.set_xlim(-1,1)\n",
    "ax.set_title('Average kmer count change vs early human samples')\n",
    "ax.set_xlabel('Average kmer count change')\n",
    "ax.set_ylabel('-log10 of p-value')\n",
    "handles, labels = plt.gca().get_legend_handles_labels()\n",
    "order = [1,0,2,3]\n",
    "plt.legend([handles[idx] for idx in order],[labels[idx] for idx in order])\n",
    "# plt.savefig(\"output/deer/15mers_change_vs_p_early.svg\", dpi = 600, format = \"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "sns.scatterplot(data=common_kmer_matrix, x='deer_late_err', y = '-log10_p-value_late',s=10, linewidth = 0, hue='group', palette=VOLCANO_CMAP)\n",
    "plt.axhline(y=-np.log10(0.01), color = 'black')\n",
    "ax.set_xlim(-1,1)\n",
    "ax.set_title('Average kmer count change vs late human samples')\n",
    "ax.set_xlabel('Average kmer count change')\n",
    "ax.set_ylabel('-log10 of p-value')\n",
    "handles, labels = plt.gca().get_legend_handles_labels()\n",
    "order = [1,0,2,3]\n",
    "plt.legend([handles[idx] for idx in order],[labels[idx] for idx in order])\n",
    "# plt.savefig(\"output/deer/change_vs_p_late.svg\", dpi = 600, format = \"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "sns.scatterplot(data = common_kmer_matrix, x='deer_early_err', y='deer_late_err',s=10, linewidth = 0, hue='group', palette = VOLCANO_CMAP)\n",
    "plt.axline((0, 0), slope=1, color=(\"black\",0.5), linestyle=\"dashed\")\n",
    "ax.set_xlim(-1,1)\n",
    "ax.set_ylim(-1,1)\n",
    "ax.set_title('Average kmer count change vs early and late human samples')\n",
    "ax.set_xlabel('Average kmer count change vs early human samples')\n",
    "ax.set_ylabel('Average kmer count change vs late human samples')\n",
    "handles, labels = plt.gca().get_legend_handles_labels()\n",
    "order = [1,0,2,3]\n",
    "plt.legend([handles[idx] for idx in order],[labels[idx] for idx in order])\n",
    "# plt.savefig(\"output/deer/change_corr.svg\", dpi = 600, format = \"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "6. Optionally re-examin rare k-mers similar to over-represented k-mers. Run the two cells below only if you want to include potential rare, but still enriched k-mers in the analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deer_overrep_rare = all_kmer_matrix.loc[(all_kmer_matrix['deer_avg'] == all_kmer_matrix[AVG_COLS].max(axis=1))]\n",
    "deer_overrep_rare = deer_overrep_rare[~deer_overrep_rare.index.isin(deer_overrep.index)]\n",
    "\n",
    "#Comment the following line to avoid lengthy k-mer similarity graph calculation\n",
    "deer_overrep_similarity_graph = portek.build_similarity_graph_two_list(deer_overrep.index.tolist(), deer_overrep_rare.index.tolist(),m)\n",
    "\n",
    "#Uncomment the first of the following blocks to save the graph as pickle file. Uncomment the second to load a previously calculated graph.\n",
    "\n",
    "# with open(f\"output/deer/dor_15mer_{m}m.pickle\", mode = \"wb\") as out_file:\n",
    "#     pickle.dump(deer_overrep_similarity_graph, out_file)\n",
    "\n",
    "# with open(f\"output/deer/dor_15mer_{m}m.pickle\", mode = \"rb\") as in_file:\n",
    "#     deer_overrep_similarity_graph = pickle.load(in_file)\n",
    "\n",
    "deer_overrep_rare = deer_overrep_rare[deer_overrep_rare.index.isin(deer_overrep_similarity_graph.nodes)]\n",
    "\n",
    "deer_overrep_rare['deer_early_err'] = deer_overrep_rare['deer_avg']-deer_overrep_rare['humearly_avg']\n",
    "deer_overrep_rare['deer_late_err'] = deer_overrep_rare['deer_avg']-deer_overrep_rare['humlate_avg']\n",
    "deer_overrep_rare['deer_RMSE'] =  np.sqrt(((deer_overrep_rare['deer_early_err'])**2+(deer_overrep_rare['deer_late_err'])**2)/2)\n",
    "deer_overrep_rare['seq'] = deer_overrep_rare.index\n",
    "deer_overrep_rare['p-value_early'] = deer_overrep_rare['seq'].apply(portek.calc_kmer_pvalue,args=(deer_sample_idx, humearly_sample_idx, deer_overrep_rare))\n",
    "deer_overrep_rare['p-value_late'] = deer_overrep_rare['seq'].apply(portek.calc_kmer_pvalue,args=(deer_sample_idx, humlate_sample_idx, deer_overrep_rare))\n",
    "deer_overrep_rare = deer_overrep_rare.sort_values('deer_RMSE',ascending=False)\n",
    "deer_overrep_rare = deer_overrep_rare.drop('seq', axis=1)\n",
    "deer_overrep_rare['-log10_p-value_early'] = -np.log10(deer_overrep_rare['p-value_early'])\n",
    "deer_overrep_rare['-log10_p-value_late'] = -np.log10(deer_overrep_rare['p-value_late'])\n",
    "deer_overrep_rare['group'] = deer_overrep_rare.apply(assign_kmer_type, axis=1)\n",
    "deer_overrep_rare  = deer_overrep_rare[deer_overrep_rare['group'] == 'deer over-represented']\n",
    "deer_overrep = pd.concat([deer_overrep_rare, deer_overrep])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_overrep_rare = all_kmer_matrix.loc[(all_kmer_matrix['deer_avg'] == all_kmer_matrix[AVG_COLS].min(axis=1))]\n",
    "human_overrep_rare = human_overrep_rare[~human_overrep_rare.index.isin(human_overrep.index)]\n",
    "\n",
    "#Comment the following line to avoid lengthy k-mer similarity graph calculation\n",
    "human_overrep_similarity_graph = portek.build_similarity_graph_two_list(human_overrep.index.tolist(), human_overrep_rare.index.tolist(),m)\n",
    "\n",
    "#Uncomment the first of the following blocks to save the graph as pickle file. Uncomment the second to load a previously calculated graph.\n",
    "\n",
    "# with open(f\"output/deer/hor_15mer_{m}m.pickle\", mode = \"wb\") as out_file:\n",
    "#     pickle.dump(human_overrep_similarity_graph, out_file)\n",
    "\n",
    "# with open(f\"output/deer/hor_15mer_{m}m.pickle\", mode = \"rb\") as in_file:\n",
    "#     human_overrep_similarity_graph = pickle.load(in_file)\n",
    "\n",
    "human_overrep_rare = human_overrep_rare[deer_overrep_rare.index.isin(deer_overrep_similarity_graph.nodes)]\n",
    "\n",
    "human_overrep_rare['deer_early_err'] = human_overrep_rare['deer_avg']-human_overrep_rare['humearly_avg']\n",
    "human_overrep_rare['deer_late_err'] = human_overrep_rare['deer_avg']-human_overrep_rare['humlate_avg']\n",
    "human_overrep_rare['deer_RMSE'] =  np.sqrt(((human_overrep_rare['deer_early_err'])**2+(human_overrep_rare['deer_late_err'])**2)/2)\n",
    "human_overrep_rare['seq'] = human_overrep_rare.index\n",
    "human_overrep_rare['p-value_early'] = human_overrep_rare['seq'].apply(portek.calc_kmer_pvalue,args=(deer_sample_idx, humearly_sample_idx, human_overrep_rare))\n",
    "human_overrep_rare['p-value_late'] = human_overrep_rare['seq'].apply(portek.calc_kmer_pvalue,args=(deer_sample_idx, humlate_sample_idx, human_overrep_rare))\n",
    "human_overrep_rare = human_overrep_rare.sort_values('deer_RMSE',ascending=False)\n",
    "human_overrep_rare = human_overrep_rare.drop('seq', axis=1)\n",
    "human_overrep_rare['-log10_p-value_early'] = -np.log10(human_overrep_rare['p-value_early'])\n",
    "human_overrep_rare['-log10_p-value_late'] = -np.log10(human_overrep_rare['p-value_late'])\n",
    "human_overrep_rare['group'] = human_overrep_rare.apply(assign_kmer_type, axis=1)\n",
    "human_overrep_rare  = human_overrep_rare[human_overrep_rare['group'] == 'human over-represented']\n",
    "human_overrep = pd.concat([human_overrep_rare, deer_overrep])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "7. Find enriched k-mers by applying the RMSE filter. Export enriched k-mers for use in host classification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deer_enriched = deer_overrep[deer_overrep['deer_RMSE']>min_RMSE]\n",
    "human_enriched = human_overrep[human_overrep['deer_RMSE']>min_RMSE]\n",
    "\n",
    "\n",
    "def assign_host_numerical(sample_id):\n",
    "    if sample_id in deer_sample_idx:\n",
    "        host = 1\n",
    "    else:\n",
    "        host = 0\n",
    "    return host\n",
    "\n",
    "\n",
    "counts_for_classifier = pd.concat([deer_enriched, human_enriched]).T\n",
    "counts_for_classifier.drop(common_kmer_stat_matrix.columns, axis=0, inplace=True)\n",
    "counts_for_classifier['host'] = counts_for_classifier.index.map(assign_host_numerical)\n",
    "\n",
    "counts_for_classifier.to_csv(\"output/deer/15mer_counts_for_classifier.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "8. Map enriched k-mers to the reference genome and identify host-enriched mutations."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
