{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This notebook descrbies the PORT-EK pipeline analysis of k-mers genreated from the \"deer\" dataset from [reference to our paper].\n",
    "To use it you will need the k-mer indices generated with PORTEKfind.py in the output/deer/15mer_indices folder.\n",
    "To generate the indices as in the paper, you will need to run PORTEKfind.py on the appropriate GISAID data sets, previously downloaded in .fasta files, using the following commands:\n",
    " - python PORTEKfind.py \"input/deer/EPI_SET_220422va.fasta\" \"output/deer/15mer_indices/\" --k 15 --group deer\n",
    " - python PORTEKfind.py \"input/deer/EPI_SET_220422rw.fasta\" \"output/deer/15mer_indices/\" --k 15 --group humearly  \n",
    " - python PORTEKfind.py \"input/deer/EPI_SET_220422qc.fasta\" \"output/deer/15mer_indices/\" --k 15 --group humlate  \n",
    "\n",
    "For the selection of optimal k please see optimal_k_selection.ipynb notebook. \n",
    "Note that GISAID web interface only allows downloading up to 10 000 sequences in a ingle file, so you may need to split the human data sets into several .fasta files. \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Import necessary libraries and PORT-EK source code:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import portek\n",
    "\n",
    "import pandas as pd\n",
    "import json\n",
    "import pathlib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Declare data set specific definitions and functions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#PORT-EK parameters\n",
    "c = 0.01  # This is the conservation thershold used in k-mer rarity filter\n",
    "m = 2  # This is the maximum number of mismatches allowed when re-examining rare k-mers\n",
    "min_RMSE = 0.1  # This is the RMSE threshold used to select enriched k-mers\n",
    "m_map = 2  # This is the maximum number of mismatches allowed when mapping k-mers to reference genome\n",
    "l_map = 1000  # This the maximum allowed offset of mapping position from average position of k-mer in samples\n",
    "\n",
    "#Relative path to k-mer indices\n",
    "INPUT_PATH = \"output/deer_early_late/15mer_indices\"\n",
    "\n",
    "#Data set specific definitions of k-mer type, column names, reference gene and protein mapping, and colors for plots.\n",
    "\n",
    "FREQ_COLS = ['deer_freq','humearly_freq','humlate_freq']\n",
    "\n",
    "VOLCANO_CMAP = {\n",
    "    \"not significant\": (\"#DDDDDD\", 0.5),\n",
    "    \"deer enriched\": (\"#ffa401\", 1),\n",
    "    \"human enriched\": (\"#005ff5\", 1),\n",
    "    \"time-dependant\": (\"#b99e9e\", 0.8),\n",
    "}\n",
    "GENE_ORDER_LIST = [\n",
    "    \"orf1ab\",\n",
    "    \"S\",\n",
    "    \"orf3a\",\n",
    "    \"E\",\n",
    "    \"M\",\n",
    "    \"orf6\",\n",
    "    \"orf7a\",\n",
    "    \"orf7b\",\n",
    "    \"orf8\",\n",
    "    \"N\",\n",
    "    \"orf10\",\n",
    "    \"intergenic\",\n",
    "]\n",
    "PROTEIN_ORDER_LIST = [\n",
    "    \"nsp1\",\n",
    "    \"nsp2\",\n",
    "    \"nsp3\",\n",
    "    \"nsp4\",\n",
    "    \"nsp5\",\n",
    "    \"nsp6\",\n",
    "    \"nsp7\",\n",
    "    \"nsp8\",\n",
    "    \"nsp9\",\n",
    "    \"nsp10\",\n",
    "    \"nsp12\",\n",
    "    \"nsp13\",\n",
    "    \"nsp14\",\n",
    "    \"nsp15\",\n",
    "    \"nsp16\",\n",
    "    \"S\",\n",
    "    \"orf3a\",\n",
    "    \"E\",\n",
    "    \"M\",\n",
    "    \"orf6\",\n",
    "    \"orf7a\",\n",
    "    \"orf7b\",\n",
    "    \"orf8\",\n",
    "    \"N\",\n",
    "    \"orf10\",\n",
    "    \"non-coding\",\n",
    "]\n",
    "\n",
    "\n",
    "def aqssign_kmer_type(row):\n",
    "    if (\n",
    "        row[\"deer_early_err\"] > 0\n",
    "        and row[\"deer_late_err\"] > 0\n",
    "        and row[\"p-value_late\"] < 0.01\n",
    "        and row[\"p-value_early\"] < 0.01\n",
    "    ):\n",
    "        return \"deer enriched\"\n",
    "    elif (\n",
    "        row[\"deer_early_err\"] < 0\n",
    "        and row[\"deer_late_err\"] < 0\n",
    "        and row[\"p-value_late\"] < 0.01\n",
    "        and row[\"p-value_early\"] < 0.01\n",
    "    ):\n",
    "        return \"human enriched\"\n",
    "    elif row[\"p-value_early\"] < 0.01 or row[\"p-value_late\"] < 0.01:\n",
    "        return \"time-dependant\"\n",
    "    else:\n",
    "        return \"not significant\"\n",
    "\n",
    "\n",
    "def assign_gene(pos):\n",
    "    if pos in range(266, 21555 + 1):\n",
    "        gene = \"orf1ab\"\n",
    "    elif pos in range(21563, 25384 + 1):\n",
    "        gene = \"S\"\n",
    "    elif pos in range(25393, 26220 + 1):\n",
    "        gene = \"orf3a\"\n",
    "    elif pos in range(26245, 26472 + 1):\n",
    "        gene = \"E\"\n",
    "    elif pos in range(26523, 27191 + 1):\n",
    "        gene = \"M\"\n",
    "    elif pos in range(27202, 27387 + 1):\n",
    "        gene = \"orf6\"\n",
    "    elif pos in range(27394, 27759 + 1):\n",
    "        gene = \"orf7a\"\n",
    "    elif pos in range(27756, 27887 + 1):\n",
    "        gene = \"orf7b\"\n",
    "    elif pos in range(27894, 28259 + 1):\n",
    "        gene = \"orf8\"\n",
    "    elif pos in range(28274, 29533 + 1):\n",
    "        gene = \"N\"\n",
    "    elif pos in range(29558, 29674 + 1):\n",
    "        gene = \"orf10\"\n",
    "    else:\n",
    "        gene = \"intergenic\"\n",
    "    return gene\n",
    "\n",
    "\n",
    "def assign_protein(pos):\n",
    "    if pos in range(266, 805 + 1):\n",
    "        protein = \"nsp1\"\n",
    "    elif pos in range(806, 2719 + 1):\n",
    "        protein = \"nsp2\"\n",
    "    elif pos in range(2720, 8554 + 1):\n",
    "        protein = \"nsp3\"\n",
    "    elif pos in range(8555, 10054 + 1):\n",
    "        protein = \"nsp4\"\n",
    "    elif pos in range(10055, 10972 + 1):\n",
    "        protein = \"nsp5\"\n",
    "    elif pos in range(10973, 11842 + 1):\n",
    "        protein = \"nsp6\"\n",
    "    elif pos in range(11843, 12091 + 1):\n",
    "        protein = \"nsp7\"\n",
    "    elif pos in range(12092, 12685 + 1):\n",
    "        protein = \"nsp8\"\n",
    "    elif pos in range(12686, 13024 + 1):\n",
    "        protein = \"nsp9\"\n",
    "    elif pos in range(13025, 13441 + 1):\n",
    "        protein = \"nsp10\"\n",
    "    elif pos in range(13442, 16236 + 1):\n",
    "        protein = \"nsp12\"\n",
    "    elif pos in range(16237, 18039 + 1):\n",
    "        protein = \"nsp13\"\n",
    "    elif pos in range(18040, 19620 + 1):\n",
    "        protein = \"nsp14\"\n",
    "    elif pos in range(19621, 20658 + 1):\n",
    "        protein = \"nsp15\"\n",
    "    elif pos in range(20659, 21555 + 1):\n",
    "        protein = \"nsp16\"\n",
    "    elif pos in range(21563, 25384 + 1):\n",
    "        protein = \"S\"\n",
    "    elif pos in range(25393, 26220 + 1):\n",
    "        protein = \"orf3a\"\n",
    "    elif pos in range(26245, 26472 + 1):\n",
    "        protein = \"E\"\n",
    "    elif pos in range(26523, 27191 + 1):\n",
    "        protein = \"M\"\n",
    "    elif pos in range(27202, 27387 + 1):\n",
    "        protein = \"orf6\"\n",
    "    elif pos in range(27394, 27759 + 1):\n",
    "        protein = \"orf7a\"\n",
    "    elif pos in range(27756, 27887 + 1):\n",
    "        protein = \"orf7b\"\n",
    "    elif pos in range(27894, 28259 + 1):\n",
    "        protein = \"orf8\"\n",
    "    elif pos in range(28274, 29533 + 1):\n",
    "        protein = \"N\"\n",
    "    elif pos in range(29558, 29674 + 1):\n",
    "        protein = \"orf10\"\n",
    "    else:\n",
    "        protein = \"non-coding\"\n",
    "    return protein"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Construct k-mer count matrix and apply rarity filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "kmer_set = set()\n",
    "sample_list = []\n",
    "in_path = pathlib.Path(INPUT_PATH).glob('**/*')\n",
    "\n",
    "for filename in in_path:\n",
    "    sample_list.append(filename.stem)\n",
    "    with open(filename, mode=\"r\") as in_file:\n",
    "        temp_dict = json.load(in_file)\n",
    "    kmer_set.update(temp_dict.keys())\n",
    "\n",
    "all_kmer_matrix = pd.DataFrame(0, index=list(kmer_set), columns=sample_list, dtype=\"uint8\")\n",
    "deer_sample_idx = [sample for sample in all_kmer_matrix.columns if \"deer\" in sample]\n",
    "humearly_sample_idx = [sample for sample in all_kmer_matrix.columns if \"humearly\" in sample]\n",
    "humlate_sample_idx = [sample for sample in all_kmer_matrix.columns if \"humlate\" in sample]\n",
    "\n",
    "print(f\"\\nImported {len(kmer_set)} kmers and {len(sample_list)} samples.\")\n",
    "\n",
    "counter = 1\n",
    "tot_files = len(sample_list)\n",
    "in_path = pathlib.Path(INPUT_PATH).glob('**/*')\n",
    "\n",
    "for filename in in_path:\n",
    "    with open(filename, mode=\"r\") as in_file:\n",
    "        temp_dict = json.load(in_file)\n",
    "    count_dict = {f\"{filename.stem}\":[len(pos) for pos in temp_dict.values()]}\n",
    "    temp_df = pd.DataFrame(count_dict, index=temp_dict.keys(), dtype=\"uint8\")\n",
    "    all_kmer_matrix.update(temp_df)\n",
    "    print(f\"Completed {filename.stem}. {counter} of {tot_files} indices done.\", end=\"\\r\", flush=True)\n",
    "    counter += 1\n",
    "\n",
    "bin_kmer_matrix = all_kmer_matrix > 0\n",
    "all_kmer_matrix['deer_freq'] = bin_kmer_matrix.loc[:,deer_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humearly_freq'] = bin_kmer_matrix.loc[:,humearly_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humlate_freq'] = bin_kmer_matrix.loc[:,humlate_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['deer_avg'] = all_kmer_matrix.loc[:,deer_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humearly_avg'] = all_kmer_matrix.loc[:,humearly_sample_idx].mean(axis=1)\n",
    "all_kmer_matrix['humlate_avg'] = all_kmer_matrix.loc[:,humlate_sample_idx].mean(axis=1)\n",
    "del bin_kmer_matrix\n",
    "\n",
    "if 'AAAAAAAAAAAAAAA' in all_kmer_matrix.index:\n",
    "    all_kmer_matrix = all_kmer_matrix.drop('AAAAAAAAAAAAAAA')\n",
    "\n",
    "common_kmer_matrix = portek.filter_kmers(all_kmer_matrix, freq_cols=FREQ_COLS, cons_thr=c)\n",
    "\n",
    "print(f\"\\n{len(common_kmer_matrix)} kmers remaining after filtering at a threshold of {c}.\")\n",
    "print(\"Finished importing count matrices and calculating frequencies.\")\n",
    "print(f\"Total matrix size is {len(kmer_set)} kmers by {len(sample_list)} samples.\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
